### PR TITLE
mv metadata to util

### PR DESCRIPTION
**Pull request by jichenjc**
metadata is widely used inside this project such as cloudprovider
and later on CSI or others, so move it to a common base will
be helpful to follow up usage

**What this PR does / why we need it**:
move metadata to util then we can reuse it later
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

```release-note
None
```

### CODE REVIEW

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ import (
 	certutil "k8s.io/client-go/util/cert"
 	cloudprovider "k8s.io/cloud-provider"
 	v1helper "k8s.io/cloud-provider-openstack/pkg/apis/core/v1/helper"
+	metadata "k8s.io/cloud-provider-openstack/pkg/util/metadata"
```
Comment (lingxiankong) [inline pkg/cloudprovider/providers/openstack/openstack.go]: The import alias is not needed.

File: pkg/util/metadata/metadata.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ func getConfigDrivePath(metadataVersion string) string {
 	return fmt.Sprintf(configDrivePathTemplate, metadataVersion)
 }
 
-func getMetadataFromConfigDrive(metadataVersion string) (*Metadata, error) {
+func GetMetadataFromConfigDrive(metadataVersion string) (*Metadata, error) {
```
Comment (lingxiankong) [inline pkg/util/metadata/metadata.go]: The method name is a little bit redundant, could we just change it to something like `getFromConfigDrive`?

File: pkg/util/metadata/metadata.go [hunk 3]
```diff
@@ -147,7 +147,7 @@ func getMetadataFromConfigDrive(metadataVersion string) (*Metadata, error) {
 	return parseMetadata(f)
 }
 
-func getMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
+func GetMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
```
Comment (lingxiankong) [inline pkg/util/metadata/metadata.go]: same here, e.g. `GetFromMetadataService`

File: pkg/util/metadata/metadata.go [hunk 4]
```diff
@@ -168,7 +168,15 @@ func getMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
...
-func getMetadata(order string) (*Metadata, error) {
+func SetMetadata(value *Metadata) {
+	metadataCache = value
+}
+
+func ClearMetadata() {
+	metadataCache = nil
+}
+
+func GetMetadata(order string) (*Metadata, error) {
```
Comment (lingxiankong) [inline pkg/util/metadata/metadata.go]: s/GetMetadata/Get

File: pkg/util/metadata/metadata.go [hunk 5]
```diff
@@ -168,7 +168,15 @@ func getMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
 // Metadata is fixed for the current host, so cache the value process-wide
 var metadataCache *Metadata
 
-func getMetadata(order string) (*Metadata, error) {
+func SetMetadata(value *Metadata) {
```
Comment (lingxiankong) [inline pkg/util/metadata/metadata.go]: s/SetMetadata/Set

File: pkg/util/metadata/metadata.go [hunk 6]
```diff
@@ -168,7 +168,15 @@ func getMetadataFromMetadataService(metadataVersion string) (*Metadata, error) {
...
+	metadataCache = value
+}
+
+func ClearMetadata() {
```
Comment (lingxiankong) [inline pkg/util/metadata/metadata.go]: s/ClearMetadata/Clear

**Comment by hogepodge** [comment]
Looks good to me, just need to address the comments

File: pkg/cloudprovider/providers/openstack/openstack.go (hunk 1 above)
Comment (jichenjc) [reply pkg/cloudprovider/providers/openstack/openstack.go]: ok

File: pkg/cloudprovider/providers/openstack/openstack_instances.go [hunk 7]
```diff
@@ -28,6 +28,7 @@ import (
 	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	cloudprovider "k8s.io/cloud-provider"
+	metadata "k8s.io/cloud-provider-openstack/pkg/util/metadata"
```
Comment (ramineni) [inline pkg/cloudprovider/providers/openstack/openstack_instances.go]: alias could be removed from here also?

File: pkg/cloudprovider/providers/openstack/openstack_test.go [hunk 8]
```diff
@@ -37,6 +37,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/rand"
 	"k8s.io/apimachinery/pkg/util/wait"
+	metadata "k8s.io/cloud-provider-openstack/pkg/util/metadata"
```
Comment (ramineni) [inline pkg/cloudprovider/providers/openstack/openstack_test.go]: ditto

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 9]
```diff
@@ -42,6 +42,7 @@ import (
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/volumeattach"
 	"github.com/prometheus/client_golang/prometheus"
 
+	metadata "k8s.io/cloud-provider-openstack/pkg/util/metadata"
```
Comment (ramineni) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: ditto

**Comment by lingxiankong** [comment]
/lgtm

thanks!

**Comment by adisky** [comment]
/lgtm
@jichenjc as a part of it you will also make CSI to use this metadata package??

**Comment by jichenjc** [comment]
@adisky yes, that's on the way :)

**Pull request merged**